Mathlib/Tactic/TermCongr.lean

```diff
@@ -0,0 +1,695 @@
+  -- Potentially metavariables get assigned as we process congruence holes,
+  -- so instantiate them to be safe. Placeholders and implicit arguments might
+  -- end up with congruence holes, so they indeed might need a nontrivial congruence.
+  let lhs ← instantiateMVars lhs
+  let rhs ← instantiateMVars rhs
+  if let some res ← mkCongrOfCHole? mvarCounterSaved lhs rhs then
+    trace[Elab.congr] "hole processing succeeded"
+    return res
+  if (hasCHole mvarCounterSaved lhs).isNone && (hasCHole mvarCounterSaved rhs).isNone then
+    -- It's safe to fastforward if the lhs and rhs are defeq and have no congruence holes.
+    -- This is more conservative than necessary since congruence holes might be inside proofs.
+    if ← isDefEq lhs rhs then
+      return {lhs, rhs, pf? := none}
+  if ← (isProof lhs <||> isProof rhs) then
+    -- We don't want to look inside proofs at all.
+    return ← CongrResult.mkDefault lhs rhs
+  match lhs, rhs with
+  | .app .., .app .. =>
+    trace[Elab.congr] "app"
+    let arity := lhs.getAppNumArgs
+    unless arity == rhs.getAppNumArgs do
+      trace[Elab.congr] "app desync (arity)"
+      return ← CongrResult.mkDefault' mvarCounterSaved lhs rhs
+    let f := lhs.getAppFn
+    let f' := rhs.getAppFn
+    unless ← isDefEq (← inferType f) (← inferType f') do
+      trace[Elab.congr] "app desync (function types)"
+      return ← CongrResult.mkDefault' mvarCounterSaved lhs rhs
+    let fnRes ← mkCongrOf (depth + 1) mvarCounterSaved f f'
+    trace[Elab.congr] "mkCongrOf functions {f}, {f'} has isRfl = {fnRes.isRfl}"
+    if !fnRes.isRfl then
+      -- If there's a nontrivial proof, then since mkHCongrWithArity fixes the function
+      -- we need to handle this ourselves.
+      let lhs := mkAppN fnRes.lhs lhs.getAppArgs
+      let lhs' := mkAppN fnRes.rhs lhs.getAppArgs
+      let rhs := mkAppN fnRes.rhs rhs.getAppArgs
+      let mut pf ← fnRes.eq
+      for arg in lhs.getAppArgs do
+        pf ← mkCongrFun pf arg
+      let res1 := CongrResult.mk' lhs lhs' pf
+      let res2 ← mkCongrOf (depth + 1) mvarCounterSaved lhs' rhs
+      return res1.trans res2
+    let thm ← mkHCongrWithArity' fnRes.lhs arity
+    let mut args := #[]
+    let mut lhsArgs := #[]
+    let mut rhsArgs := #[]
+    let mut nontriv : Bool := false
+    for lhs' in lhs.getAppArgs, rhs' in rhs.getAppArgs, kind in thm.argKinds do
+      match kind with
+      | .eq =>
+        let res ← mkCongrOf (depth + 1) mvarCounterSaved lhs' rhs'
+        nontriv := nontriv || !res.isRfl
+        args := args |>.push res.lhs |>.push res.rhs |>.push (← res.eq)
+        lhsArgs := lhsArgs.push res.lhs
+        rhsArgs := rhsArgs.push res.rhs
+      | .heq =>
+        let res ← mkCongrOf (depth + 1) mvarCounterSaved lhs' rhs'
+        nontriv := nontriv || !res.isRfl
+        args := args |>.push res.lhs |>.push res.rhs |>.push (← res.heq)
+        lhsArgs := lhsArgs.push res.lhs
+        rhsArgs := rhsArgs.push res.rhs
```

Comment:
I guess you could remove the duplication here by writing a single function `res.or_or_heq kind` or something. Probably doesn't matter, but it would have saved my taking a moment to work out if anything was different.

Reply:
Yeah, I'd considered trying to deduplicate, but what stopped me before is all these `mut` variables

Reply:
Ah, indeed. Okay.